---
title: "ggtree: a phylogenetic tree viewer for different types of tree annotations"
author: "Guangchuang Yu <br />
        <guangchuangyu@gmail.com> <br />
        School of Public Health <br />
	Li Ka Shing Faculty of Medicine <br />
        The University of Hong Kong"
date: "`r Sys.Date()`"
output: BiocStyle::html_document
bibliography: ggtree.bib
csl: nature.csl
vignette: >
  %\VignetteIndexEntry{ggtree: phylogenetic tree viewer and annotator}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r style, echo=FALSE, results="asis"}
BiocStyle::markdown()
```


> You can't even begin to understand biology, you can't understand life, unless you understand what it's all there for, how it arose - and that means evolution.
> --- Richard Dawkins


# Introduction
This project came out with my needs to annotate nucleotide substitutions in the phylogenetic tree, and I found that there is no tree visualization software can do this easily. All the tree viewer are designed for displaying phylogenetic tree, but not annotating it. Although some tree viewers support displaying bootstrap values in the tree, it is hard/impossible to display other information in the tree. My first solution for displaying nucleotide substituition in the tree is to add these information in the node/tip names and use traditional tree viewer to show it. I displayed the information in the tree successfully, but I do believe this quick-and-dirty hack is ugly.

In the old day, we even don't have enough sequences to infer phylogenetic tree. At that time, as we almost don't have a need to annotate a tree, displaying the evolution relationships is mostly enough. Nowaday, we can obtain a lot of data from different experiments, and we want to associate our data, for instance antigenic change, with the evolution relationship. Visualizing these associations in the phylogenetic tree can help us to identify evolution patterns. I believe we need a next generation tree viewer that can view a phylogenetic tree easily as we did with classical software and support adding annotation data in a layer one by one above the tree. This is the objective of developing the `r Githubpkg("GuangchuangYu/ggtree")`. Common tasks of annotating a phylogenetic tree should be easy and complicated tasks can be possible to achieve by adding multiple layers of annotation.

The `r Githubpkg("GuangchuangYu/ggtree")` is designed by extending the `r CRANpkg("ggplot2")`[@wickham_ggplot2_2009] package. It based on grammar of graphics and takes all the good parts of `r CRANpkg("ggplot2")`. There are other R packages that implemented tree viewer using `r CRANpkg("ggplot2")`, including `r CRANpkg("OutbreakTools")`, `r Biocpkg("phyloseq")`[@mcmurdie_phyloseq_2013] and `r Githubpkg("gjuggler/ggphylo")`, but all of them only create complex tree view function for their specific needs. They are just classical tree viewers that only viewing the tree or annotating a specific data type. The good parts of `r CRANpkg("ggplot2")` are not available in these packages. They lack of flexibilities of annotating phylogenetic tree by diverse user inputs. 

# Phylogenetic tree visualization
## Viewing tree with `ggtree`
`r Githubpkg("GuangchuangYu/ggtree")` extend `ggplot` to support viewing phylogenetic tree. It implement `geom_tree` layer for displaying phylogenetic tree, as shown below:
```{r fig.width=4, fig.height=4, tidy=TRUE, width=60}
nwk <- system.file("extdata", "sample.nwk", package="ggtree")
print(readLines(nwk))
library("ggtree")

tree <- read.tree(nwk)
ggplot(tree, aes(x, y)) + geom_tree() + theme_tree() + xlab("") + ylab("")
```

This example tree was obtained from Chapter 34 of "Inferring Phylogenies"[@felsenstein_inferring_2003 pp.573]. The function, `ggtree`, was implemented as a short cut to visualize a tree, and it works exactly the same as shown above.

`ggtree` takes all the advantages of `ggplot2`. For example, we can change the color, size and type of the lines as we did with `ggplot2`.
```{r fig.width=4, fig.height=4}
ggtree(tree, color="steelblue", size=0.5, linetype="dotted")
```

By default, the tree is viewing in ladderize form, user can set the parameter `ladderize = FALSE` to disable it.
```{r fig.width=4, fig.height=4}
ggtree(tree, ladderize=FALSE)
```

The `branch.length` is used to scale the edge, user can set the parameter `branch.length = "none"` to only viewing the tree topology.

```{r fig.width=4, fig.height=4}
ggtree(tree, branch.length="none")
```

## Layout

Currently, `ggtree` supports several layout, including `phylogram` (by default), `cladogram`, `dendrogram`, `fan`, and `unrooted`. Unrooted layout was implemented by the `equal-angle algorithm` that described in "Inferring Phylogenies"[@felsenstein_inferring_2003 pp.578-580].

```{r fig.width=8, fig.height=12}
library("gridExtra")
grid.arrange(ggtree(tree) + ggtitle("phylogram layout"),
	     ggtree(tree, layout="dendrogram") + ggtitle("dendrogram layout"),
	     ggtree(tree, layout="cladogram") + ggtitle("cladogram layout"),
	     ggtree(tree, layout="cladogram", branch.length="none") + scale_x_reverse()+coord_flip() + ggtitle("cladogram layout"),
	     ggtree(tree, layout="fan") + ggtitle("fan layout"),
	     ggtree(tree, layout="unrooted") + ggtitle("unrooted layout"),
	     ncol=2)
```


## Support multiple phylogenetic classes

`r Githubpkg("GuangchuangYu/ggtree")` defined several S4 classes to store phylogenetic object and its associated annotation, including `jplace`, `palm_rst`, `codeml_mlc`, `codeml`, `hyphy` and `beast`. In addition, it also supports `phylo` (defined by `r CRANpkg("ape")`[@paradis_ape_2004]), and `phylo4` (defined by `r CRANpkg("phylobase")`)

```{r fig.width=12, fig.height=4}
print(tree)
p1 <- ggtree(tree) + ggtitle("tree of phylo object")
library(phylobase)
tr2 <- as(tree, "phylo4")
str(tr2)
p2 <- ggtree(tr2) + ggtitle("tree of phylo4 object")
jpf <- system.file("extdata", "sample.jplace", package="ggtree")
jp <- read.jplace(jpf)
print(jp)
p3 <- ggtree(jp) + ggtitle("tree of jplace object")
library(gridExtra)
grid.arrange(p1, p2, p3, ncol=3)
```
 
## Display evolution distance

To show evolution distance, we can use `theme_tree2()` or `ggtree(showDistance=TRUE)`

```{r fig.width=6}
ggtree(tree) + theme_tree2()
```

Another way is to show the edge length of the tree: 
```{r fig.width=6, warning=FALSE}
ggtree(tree, showDistance=TRUE) +geom_text(aes(label=length), hjust=1, vjust=-0.5, color="#F06C45")
```

The scale of branch length can be specify via `scale_x_continuous()`.
```{r fig.width=6}
ggtree(tree) + theme_tree2() + scale_x_continuous(breaks=seq(0, 60, 5))
```
 
## Display nodes/tips

Show all the internal nodes and tips in the tree can be done by adding a layer of points using `geom_point`.

```{r fig.width=6}
ggtree(tree)+geom_point(aes(shape=isTip, color=isTip), size=3)
```
 
And of course, we can separate nodes and tips by using `subset`.
```{r fig.width=6}
p <- ggtree(tree) + geom_point(subset=.(!isTip), 
     		       	       color="#b5e521", alpha=1/4, size=10)
print(p)
```

```{r fig.width=6}
p + geom_point(color="#FDAC4F", shape=8, size=3, subset=.(isTip))
```

## Display labels
```{r fig.width=6, warning=FALSE}
p + geom_text(aes(label=label), size=6, color="purple", hjust=-0.3)
```

By default, the positions are based on the node positions, we can change them to based on the middle of the branch/edge.

```{r fig.width=6, fig.height=6, warning=FALSE}
p + geom_text(aes(x=branch, label=label), size=6, color="purple", vjust=-0.3)
```

Based on the middle of branches is very useful when annotating transition from parent node to child node, please refer to [ggtree-treeAnnotation](ggtree-treeAnnotation.html).

## Theme
`theme_tree()` defined a totally blank canvas, while `theme_tree2()` add phylogenetic distance legend. These two themes all accept a parameter of `bgcolor` that defined the background color.

```{r fig.width=12, fig.height=5}
grid.arrange(
	ggtree(rtree(30), color="red") + theme_tree("steelblue"),
	ggtree(rtree(20), color="white") + theme_tree("black"),
	ncol=2)
```

## Update tree viewing with a new tree
In the __Display nodes/tips__ section, we have a `p` object that stored the tree viewing of 13 tips and internal nodes highlighted with specific colored big dots. If we want to applied this pattern, and we can imaging more complex one, to a new tree, we don't need to build the tree step by step. `r Githubpkg("GuangchuangYu/ggtree")` provides an operator, `%<%`, for applying the visualization pattern to a new tree.

For example, the pattern in the `p` object will be applied to a new tree with 50 tips as shown below:
```{r fig.width=6, fig.height=5}
p %<% rtree(50)
```


# Tree annotation
In `r Githubpkg("GuangchuangYu/ggtree")`, we implemented several functions to parse the output from [PAML](http://abacus.gene.ucl.ac.uk/software/paml.html)[@yang_paml_2007], [HYPHY](http://hyphy.org/w/index.php/Main_Page)[@pond_hyphy_2005], [EPA](http://sco.h-its.org/exelixis/web/software/epa/index.html)[@berger_EPA_2011], [PPLACER](http://matsen.fhcrc.org/pplacer/)[@matsen_pplacer_2010] and [BEAST](http://beast2.org/)[@bouckaert_beast_2014] and defined several classes to store phylogenetic object and associated annotation, including `palm_rst`, `codeml_mlc`, `baseml`, `codeml`, `hyphy`, `jplace`, and `beast`.


## annotating tree with BEAST output
```{r}
file <- system.file("extdata/BEAST", "beast_mcc.tree", package="ggtree")
beast <- read.beast(file)
beast
```
Since `%` is not a valid character in `names`, all the feature names that contain `%` will convert to `.`. For example, `height_95%_HPD` will be changed to `height_95._HPD`.


```{r fig.width=8, tidy=TRUE, width=60, warning=FALSE}
plot(beast, annotation="height")
```

User can round the digits by setting the parameter `ndigits`.
```{r fig.width=8, tidy=TRUE, width=60, warning=FALSE}
plot(beast, annotation="height", ndigits=2)
```


```{r fig.width=8, tidy=TRUE, width=60, warning=FALSE}
plot(beast, annotation="length_95._HPD", ndigits=2, branch.length="none") + theme_tree()
```

## annotating tree with PAML output
### BASEML
#### rst file
`rst` file from `baseml` is similar to `codeml` output. The only difference is the space in the sequences. For `baseml`, each ten bases are separated by one space, while for `codeml`, each three bases (triplet) are separated by one space. We defined a `read.paml_rst` to parse `rst` file. It support `baseml` and `codeml` output. The information will be stored in `paml_rst` S4 object.

```{r}
rstfile <- system.file("extdata/PAML_Baseml", "rst", package="ggtree")
tipfas <- system.file("extdata", "pa.fas", package="ggtree")
rst <- read.paml_rst(rstfile, tipfas)
rst
```

The function `read.paml_rst` can only accept one parameter, `rstfile`, and the output can be used to view the phylogeny. But if we want to view the substitution annotation, we should provide tip sequence fasta file to parameter `tip.fasfile`, since `rstfile` only contain inferred ancestral sequences.

`mlb` file of `baseml` output do contain tip sequences, and `r Githubpkg("GuangchuangYu/ggtree")` provides another function `read.baseml` to parse `rstfile` and `mlbfile` simultaneously.

So for viewing substitution annotation, we can use `read.paml_rst` with `rstfile` and `tip.fasfile` provided or use `read.baseml` with `rstfile` and `mlbfile` provided.

```{r}
mlbfile <- system.file("extdata/PAML_Baseml", "mlb", package="ggtree")
baseml <- read.baseml(rstfile, mlbfile)
baseml
```

```{r fig.width=10, fig.height=8, tidy=TRUE, width=60, warning=FALSE}
plot(rst, annotation="marginal_AA_subs")
```

### CODEML
#### rst file
```{r fig.width=10, fig.height=8, tidy=TRUE, width=60, warning=FALSE}
rstfile <- system.file("extdata/PAML_Codeml", "rst", package="ggtree")
rst <- read.paml_rst(rstfile, tip.fasfile=tipfas)
rst
plot(rst, annotation="joint_AA_subs")
```

#### mlc file
```{r fig.width=8, tidy=TRUE, width=60, warning=FALSE}
mlcfile <- system.file("extdata/PAML_Codeml", "mlc", package="ggtree")
mlc <- read.codeml_mlc(mlcfile)
mlc
plot(mlc, branch.length="branch.length", annotation="dN/dS", ndigits=2)
```

The paramter `branch.length` can be one of available annotations:
```{r}
get.fields(mlc)
```

For example, if we set `branch.length` to `dN/dS`, it will plot the $\omega$ (`dN/dS`) tree:
```{r fig.width=8, tidy=TRUE, width=60, warning=FALSE}
plot(mlc, branch.length="dN/dS", annotation="dN/dS", ndigits=2)
```

We can also plot the `dN` or `dS` tree and others. The parameter `annotation` can also be one of the available annotations.


#### mlc and rst files

We can annotate the tree with information presented in `rstfile` and `mlcfile` separately as demonstrated in previous sessions.

We recommend using them both, and user don't need to provide tip sequences, as it's already available in `mlcfile`. All the features in both files are available for annotation.


```{r fig.width=12, fig.height=8, tidy=TRUE, width=60, warning=FALSE}
ml <- read.codeml(rstfile, mlcfile)
ml
plot(ml, branch.length="rst.branch.length", annotation="dN/dS")
plot(ml, branch.length="mlc.branch.length", annotation="joint_AA_subs")
plot(ml, branch.length="dN/dS", annotation="joint_AA_subs")
```

## annotating tree with HYPHY output
```{r}
nwk <- system.file("extdata/HYPHY", "labelledtree.tree", package="ggtree")
ancseq <- system.file("extdata/HYPHY", "ancseq.nex", package="ggtree")
hy <- read.hyphy(nwk, ancseq, tipfas)
hy
```

```{r fig.width=12, fig.height=10, tidy=TRUE, width=60, warning=FALSE}
plot(hy, annotation="AA_subs")
```


## annotating tree with EPA and PPLACER output
[EPA](http://sco.h-its.org/exelixis/web/software/epa/index.html)[@berger_EPA_2011] and [PPLACER](http://matsen.fhcrc.org/pplacer/)[@matsen_pplacer_2010] have common output file format, `jplace`. 



### user specific annotation
We provides several functions to parse and store information from common software output in this field, and corresponding `plot` methods for visualizing annotation in the tree.

Here, we would like to demonstrate how to inject user specific annotation data in a tree.

Suppose we have the following data that associated with the tree and would like to attach the data in the tree.

```{r}
nwk <- system.file("extdata", "sample.nwk", package="ggtree")
tree <- read.tree(nwk)
p <- ggtree(tree)

dd <- data.frame(taxa=LETTERS[1:13], 
      		 place=c(rep("GZ", 5), rep("HK", 3), rep("CZ", 4), NA),
                 value=round(abs(rnorm(13, mean=70, sd=10)), digits=1))
## you don't need to order the data
## data was reshuffled just for demonstration
dd <- dd[sample(1:13, 13), ]
row.names(dd) <- NULL
```
```{r eval=FALSE}
print(dd)
```

```{r echo=FALSE, results='asis'}
knitr::kable(dd)
```

We can imaging that the `place` column is the place we isolated the species and `value` column stored numerical values for example bootstrap values.

We have shown using the operator, `%<%`, to update a tree view with a new tree. Here, we will introduce another operator, `%<+%`, that attach annotation data to a tree view. The only requirement of the input data is that its first column should be matched with the node labels of the tree.

After attaching the annotation data to the tree by `%<+%`, all the columns in the data are visible to `ggplot2`. As an example, here we attach the above annotation data to the tree view, `p`, and add a layer that showing the tip labels and colored by the isolation site stored in `place` column.

```{r fig.width=6, fig.height=5, warning=FALSE}
p <- p %<+% dd + geom_text(aes(color=place, label=label), hjust=-0.5)
print(p)
```

We can add another layer to display the isolation sites.
```{r fig.width=6, fig.height=5, warning=FALSE}
p <- p + geom_text(aes(color=place, label=place), hjust=1, vjust=-0.4, size=3)
print(p)
```

And another layer showing numerical values:
```{r fig.width=6, fig.height=5, warning=FALSE}
p <- p + geom_text(aes(color=place, label=value), hjust=1, vjust=1.4, size=3)
print(p)
```

```{r fig.width=7, fig.height=5, warning=FALSE}
p <- p + geom_point(aes(size=value, shape=place, color=place), alpha=0.25, subset=.(isTip))
print(p)
```

## `jplace` file format
The `jplace` file format was defined by Masten[@matsen_format_2012] for phylogenetic placements. We employ this file format to store user specific annotation. Suppose we have a tree, and the associated data that indicate the amino acid substitutions from parent node to child node as shown below:
```{r}
tree <- system.file("extdata", "pa.nwk", package="ggtree")
data <- read.csv(system.file("extdata", "pa_subs.csv", package="ggtree"), stringsAsFactor=FALSE)
print(tree)
head(data)
```

`r Githubpkg("GuangchuangYu/ggtree")` provides a function, `write.jplace`, to combine a tree and an associated data and store them to a single `jplace` file.
```{r eval=FALSE}
write.jplace(tree, data, "pa.jplace")
```

Then user can use `read.jplace` function to parse the `jplace` file and store the information to a `jplace` object.
```{r}
jpf <- system.file("extdata", "pa.jplace", package="ggtree")
jp <- read.jplace(jpf)
print(jp)
```

The variables of the associated annotation data can be access via `get.fields`.
```{r}
get.fields(jp)
```
Now we know the `jp` object stored the tree and the associated amino acid substitution information, we can view the tree and display the associated annotation data on it directly by `ggtree`.

```{r fig.width=12, fig.height=12, warning=FALSE}
ggtree(jp, showDistance=TRUE) + 
	   geom_text(aes(x=branch, label=subs), color="purple", vjust=-1, size=3) + 
	   geom_text(aes(label=label), hjust=-.5)
```

## Visualize tree and associated matrix
```{r}
nwk <- system.file("extdata", "sample.nwk", package = "ggtree")
require(ape)
tree <- read.tree(nwk)

d <- matrix(sample(1:4, 13*4, replace=T), ncol=4)
row.names(d) <-  tree$tip.label
colnames(d) <- paste0("g", 1:4)
```
```{r eval=FALSE}
print(d)
```

```{r echo=FALSE}
knitr::kable(d)
```

```{r fig.width=12, fig.heigh=6}
gplot(tree, d, low="green", high="red", widths=c(.7, .3))
```

```{r}
d <- matrix(abs(rnorm(13*5)), ncol=5)
row.names(d) <- tree$tip.label
colnames(d) <- paste0("g", 1:5)
```
```{r eval=FALSE}
print(d)
```

```{r echo=FALSE}
knitr::kable(d)
```

```{r fig.width=12, fig.heigh=6}
gplot(tree, d, low="green", high="red", widths=c(.6, .4))
```




# Session info
Here is the output of `sessionInfo()` on the system on which this document was compiled:
```{r echo=FALSE}
sessionInfo()
```


# References